delegate-app/Main.hs

```diff
@@ -1,62 +1,336 @@
+  broadcast = forever $ do
+    flushAndTraverseNewElements
+      eventQueue
+      ( delegateEventStep
+          -- All delegateEventStep responses should be brodcasted
+          >=> return . fmap (Broadcast,)
+          >=> putInToClientsQueue
+      )
+    flushAndTraverseNewElements
+      frontendRequestQueue
+      ( delegateFrontendRequestStep
+          >=> putInToClientsQueue
+      )
+    -- FIXME: log queues overload and make tick not-static
+    liftIO $ threadDelay tick
+    where
+      -- TODO: this is not actually broadcast now
+      putInToClientsQueue responses = do
+        liftIO . atomically $
+          traverse_ (writeTChan broadcast) responses
+      -- TODO: make work
+      -- lift $ traverse_ (trace . DelegateOutput) responses
+      flushAndTraverseNewElements queue handler = do
+        elements <- liftIO . atomically $ flushTQueue queue
+        traverse_ handler elements
+
+-- | run a delegate server
+runDelegateServer ::
+  -- | the configuration of the delegate server
+  DelegateServerConfig ->
+  DelegateTracerT IO ()
+runDelegateServer conf = do
+  trace (Started $ port conf)
+
+  tracer <- askTracer
+
+  eventQueue <- liftIO . atomically $ do
+    q <- newTQueue
+    writeTQueue q Start
+    pure q
+  frontendRequestQueue <- liftIO . atomically $ newTQueue
+  toClientsChannel <- liftIO . atomically $ newTChan
+
+  clientCounter <- liftIO $ newMVar 0
+
+  let workerAction threadSort =
+        withStartStopTrace threadSort $
+          case threadSort of
+            WebsocketThread -> do
+              liftIO $
+                runServer (show $ host conf) (fromIntegral $ port conf) $
+                  runWithTracer' tracer
+                    . websocketsServer
+                      (ping conf)
+                      (freshClientIdGenerator clientCounter)
+                      frontendRequestQueue
+                      toClientsChannel
+            DelegateRunnerThread ->
+              execDelegateRunnerT $
+                runDelegateLogicSteps
+                  (tick conf)
+                  eventQueue
+                  frontendRequestQueue
+                  toClientsChannel
+            QueueAuctionStageThread ->
+              mbQueueAuctionPhases eventQueue toClientsChannel
 
-consumer :: Chan DelegateInput -> DelegateRunnerT IO ()
-consumer delegateInputs = do
-  -- This will block until an event appears in the queue
-  delegateIn <- liftIO $ readChan delegateInputs
-  delegateResponse <- delegateStep delegateIn
-  liftIO $ mbQueueAuctionPhases delegateResponse delegateInputs
-  -- FIXME: send response to client
   liftIO $
-    putStrLn $
-      "Delegate responses for input: " <> show delegateResponse
+    mapConcurrently_
+      (runWithTracer' tracer . workerAction)
+      [ WebsocketThread
+      , QueueAuctionStageThread
+      , DelegateRunnerThread
+      ]
+  where
+    freshClientIdGenerator clientCounter = do
+      v <- takeMVar clientCounter
+      putMVar clientCounter (v + 1)
```

Comment:
What is the difference? This changes atomicity somehow?

Reply:
Im curious about this as well, what scenario are you thinking of @MangoIV?

Reply:
it's not so important, but if we take out the MVar and expect to get the same next value in the counter, then that's an incorrect assumption because it will be the next next value.

Reply:
> and expect to get the same next value in the counter

Why would somebody except that?